File: offchain/test-hydra.ts
/**
 * Quick Hydra Connection Test
 * Run with: npx ts-node test-hydra.ts
 */

const WebSocket = require('ws');

const HYDRA_URL = 'ws://127.0.0.1:4001';

console.log('🧪 Testing Hydra Connection...');
console.log(`   URL: ${HYDRA_URL}`);
console.log('');

const ws = new WebSocket(HYDRA_URL);

ws.on('open', () => {
  console.log('✅ Connected to Hydra node!');
});

ws.on('message', (data: Buffer) => {
  try {
    const message = JSON.parse(data.toString());
    console.log(`📨 Received: ${message.tag}`);
    
    if (message.tag === 'Greetings') {
      console.log('');
      console.log('🤝 Hydra Greetings:');
      console.log(`   Head Status: ${message.headStatus?.tag || message.headStatus || 'Unknown'}`);
      console.log(`   Snapshot #: ${message.snapshotNumber || 0}`);
      console.log(`   Head ID: ${message.headId || 'None'}`);
      console.log(`   UTxOs: ${Object.keys(message.utxo || {}).length}`);
      console.log('');
      
      // If head is idle, we can init
      if (message.headStatus?.tag === 'Idle' || message.headStatus === 'Idle') {
        console.log('💡 Head is idle. You can initialize with:');
        console.log('   ws.send(JSON.stringify({ tag: "Init" }))');
      }
      
      // Close after getting greeting
      setTimeout(() => {
        console.log('');
        console.log('✅ Test complete! Hydra node is working.');
        ws.close();
        process.exit(0);
      }, 1000);
    }
  } catch (e) {
    console.error('Failed to parse message:', e);
  }
});

ws.on('error', (error: Error) => {
  console.error('❌ Connection error:', error.message);
  process.exit(1);
});

ws.on('close', () => {
  console.log('Connection closed');
});

// Timeout after 10 seconds
setTimeout(() => {
  console.error('❌ Timeout - no response from Hydra node');
  process.exit(1);
}, 10000);

File: test_hydra.py
#!/usr/bin/env python3
"""Quick Hydra connection test"""
import asyncio
import json

try:
    import websockets
except ImportError:
    print("Installing websockets...")
    import subprocess
    subprocess.run(["pip3", "install", "websockets", "-q"])
    import websockets

async def test_hydra():
    url = "ws://127.0.0.1:4001"
    print(f"🔌 Connecting to {url}...")
    
    try:
        async with websockets.connect(url) as ws:
            print("✅ Connected to Hydra!")
            
            # Wait for Greetings message
            msg = await asyncio.wait_for(ws.recv(), timeout=5)
            data = json.loads(msg)
            
            print(f"\n📨 Received: {data.get('tag', 'unknown')}")
            
            if data.get('tag') == 'Greetings':
                status = data.get('headStatus', {})
                if isinstance(status, dict):
                    status = status.get('tag', 'unknown')
                print(f"   Head Status: {status}")
                print(f"   Snapshot #: {data.get('snapshotNumber', 0)}")
                
                utxos = data.get('utxo', {})
                print(f"   UTxOs in Head: {len(utxos)}")
                
                if utxos:
                    print("\n   UTxO Summary:")
                    for txin, output in list(utxos.items())[:3]:
                        addr = output.get('address', 'unknown')[:30]
                        value = output.get('value', {})
                        lovelace = value.get('lovelace', value) if isinstance(value, dict) else value
                        print(f"     - {txin[:20]}... → {lovelace} lovelace")
            
            print("\n✅ Hydra node is working correctly!")
            
    except asyncio.TimeoutError:
        print("❌ Timeout waiting for response")
    except ConnectionRefusedError:
        print("❌ Connection refused - is Hydra running?")
    except Exception as e:
        print(f"❌ Error: {e}")

if __name__ == "__main__":
    asyncio.run(test_hydra())
